File: apps/knowledge/scripts/generate-article.ts
import { parseArgs } from "node:util";
import path from "node:path";
import fs from "node:fs";
import {
  CONTENT_DIR,
  getCategory,
  getExistingArticles,
  getMissingSlugs,
  getRelatedTopics,
  getStyleSample,
} from "../lib/article-catalog";
import { runArticleGenerate } from "../src/lib/langgraph-client";

interface Args {
  slug?: string;
  topic?: string;
  model?: string;
  dryRun: boolean;
  batch: boolean;
  listMissing: boolean;
  update: boolean;
  graph: boolean;
}

function parseCliArgs(): Args {
  const { values, positionals } = parseArgs({
    args: process.argv.slice(2),
    options: {
      topic: { type: "string" },
      model: { type: "string" },
      "dry-run": { type: "boolean", default: false },
      batch: { type: "boolean", default: false },
      "list-missing": { type: "boolean", default: false },
      update: { type: "boolean", default: false },
      graph: { type: "boolean", default: false },
    },
    allowPositionals: true,
  });
  return {
    slug: positionals[0],
    topic: values.topic,
    model: values.model,
    dryRun: Boolean(values["dry-run"]),
    batch: Boolean(values.batch),
    listMissing: Boolean(values["list-missing"]),
    update: Boolean(values.update),
    graph: Boolean(values.graph),
  };
}

async function runSingle(
  slug: string,
  topic: string,
  dryRun: boolean,
): Promise<{ wordCount: number; revisions: number }> {
  const result = await runArticleGenerate({
    slug,
    topic,
    category: getCategory(slug),
    relatedTopics: getRelatedTopics(slug),
    existingArticles: getExistingArticles(),
    styleSample: getStyleSample(),
  });
  if (!dryRun) {
    const outPath = path.join(CONTENT_DIR, `${slug}.md`);
    fs.writeFileSync(outPath, result.final);
  }
  return { wordCount: result.word_count, revisions: result.revisions };
}

async function runBatch(dryRun: boolean): Promise<void> {
  const missing = getMissingSlugs();
  if (missing.length === 0) {
    console.log("All articles already exist!");
    return;
  }
  console.log(`Missing articles: ${missing.length}`);
  missing.forEach((s) => console.log(`  - ${s}`));
  console.log();

  for (let i = 0; i < missing.length; i++) {
    const slug = missing[i];
    const topic = slug.replace(/-/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());
    console.log(`[${i + 1}/${missing.length}] Generating: ${slug}`);
    const t0 = Date.now();
    try {
      const r = await runSingle(slug, topic, dryRun);
      const elapsed = ((Date.now() - t0) / 1000).toFixed(0);
      console.log(`  -> ${r.wordCount} words in ${elapsed}s\n`);
    } catch (err) {
      console.log(`  -> FAILED: ${(err as Error).message}\n`);
    }
  }
}

function printGraph() {
  console.log("```mermaid");
  console.log("graph TD");
  console.log("    START((Start)) --> research");
  console.log("    research --> outline");
  console.log("    outline --> draft");
  console.log("    draft --> review");
  console.log("    review --> revise_loop{quality OK?}");
  console.log('    revise_loop -->|"yes"| finalize');
  console.log('    revise_loop -->|"no, <2 revs"| revise');
  console.log("    revise --> revise_loop");
  console.log("    finalize --> END((End))");
  console.log("```");
}

async function main() {
  const args = parseCliArgs();

  if (args.model) process.env.LLM_MODEL = args.model;

  if (args.graph) {
    printGraph();
    return;
  }

  if (args.listMissing) {
    const missing = getMissingSlugs();
    if (missing.length === 0) {
      console.log("All articles exist!");
    } else {
      console.log(`Missing articles (${missing.length}):`);
      missing.forEach((s) => console.log(`  ${s}`));
    }
    return;
  }

  if (args.batch) {
    await runBatch(args.dryRun);
    return;
  }

  const slug = args.slug;
  if (!slug) {
    console.error("Error: slug is required (or use --batch / --list-missing)");
    process.exit(1);
  }

  const existingFile = path.join(CONTENT_DIR, `${slug}.md`);
  if (fs.existsSync(existingFile) && !args.dryRun && !args.update) {
    console.log(`Article already exists: ${existingFile}`);
    console.log("Use --update to regenerate, or --dry-run to preview.");
    return;
  }

  const topic =
    args.topic ??
    slug.replace(/-/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());
  const model = process.env.LLM_MODEL ?? "deepseek-chat";
  const action = fs.existsSync(existingFile) ? "Updating" : "Generating";
  console.log(`${action}: ${topic} (${slug})`);
  console.log(`Model:      ${model}`);
  console.log(
    `Pipeline:   research -> outline -> draft -> review -> [revise loop] -> finalize`,
  );
  console.log();

  const t0 = Date.now();
  const r = await runSingle(slug, topic, args.dryRun);
  const elapsed = ((Date.now() - t0) / 1000).toFixed(0);
  console.log(
    `\nDone! ${r.wordCount} words, ${r.revisions} revisions, ${elapsed}s`,
  );
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: apps/knowledge/backend/scripts/generate_article.py
"""Pure-Python knowledge-base article generator.

Invokes the LangGraph `article_generate` StateGraph directly (no HTTP, no TS
client). The graph runs research → outline → draft → review → revise (loop) →
finalize, with DeepSeek as the per-node LLM.

Usage (run from `apps/knowledge/backend/`):
    ./run-with-env.sh ../.env.local ../../../.env -- \\
      .venv/bin/python -m scripts.generate_article \\
        --slug postgresql-joins \\
        --topic "PostgreSQL JOINs: Inner, Outer, Cross, Self, Lateral & Performance" \\
        [--category "Software Engineering"] \\
        [--related "indexing,query-optimization"]

Catalog context (existing articles, style sample) is derived by scanning the
`apps/knowledge/content/` directory directly — no Next.js / TypeScript glue.

Writes the final markdown to `apps/knowledge/content/<slug>.md`.
"""

from __future__ import annotations

import argparse
import asyncio
import sys
from pathlib import Path

# Make `knowledge_agent` importable when invoked as `python -m scripts.generate_article`
_BACKEND_DIR = Path(__file__).resolve().parent.parent
if str(_BACKEND_DIR) not in sys.path:
    sys.path.insert(0, str(_BACKEND_DIR))

from knowledge_agent.article_generate_graph import build_graph  # noqa: E402

CONTENT_DIR = _BACKEND_DIR.parent / "content"
STYLE_SAMPLE_MAX_CHARS = 2000
EXISTING_ARTICLES_LIMIT = 40


def gather_existing_articles(current_slug: str) -> str:
    items: list[str] = []
    for path in sorted(CONTENT_DIR.glob("*.md")):
        if path.stem == current_slug:
            continue
        title = path.stem.replace("-", " ").title()
        items.append(f"- [{title}](/{path.stem})")
        if len(items) >= EXISTING_ARTICLES_LIMIT:
            break
    return "\n".join(items)


def pick_style_sample(current_slug: str) -> str:
    """Use the longest existing article (other than current) as style reference."""
    candidates = [p for p in CONTENT_DIR.glob("*.md") if p.stem != current_slug]
    if not candidates:
        return ""
    best = max(candidates, key=lambda p: p.stat().st_size)
    return best.read_text(encoding="utf-8")[:STYLE_SAMPLE_MAX_CHARS]


def humanize_slug(slug: str) -> str:
    return slug.replace("-", " ").replace("_", " ").title()


async def run(
    slug: str,
    topic: str,
    category: str,
    related: str,
) -> dict:
    state = {
        "slug": slug,
        "topic": topic,
        "category": category,
        "related_topics": related,
        "existing_articles": gather_existing_articles(slug),
        "style_sample": pick_style_sample(slug),
    }
    graph = build_graph()
    result = await graph.ainvoke(state)
    return result


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    ap.add_argument("--slug", required=True, help="Article slug, e.g. postgresql-joins")
    ap.add_argument("--topic", default=None, help="Human-readable topic; defaults to titled slug")
    ap.add_argument("--category", default="", help="Lesson category (free-form)")
    ap.add_argument("--related", default="", help="Comma-separated related slugs")
    ap.add_argument("--no-write", action="store_true", help="Print but don't overwrite the .md")
    args = ap.parse_args()

    topic = args.topic or humanize_slug(args.slug)
    out_path = CONTENT_DIR / f"{args.slug}.md"
    print(f"Generating: {topic}  (slug={args.slug})")
    print(f"Output:     {out_path}")
    print("Pipeline:   research -> outline -> draft -> review -> [revise loop] -> finalize")
    print()

    result = asyncio.run(run(args.slug, topic, args.category, args.related))

    final = result.get("final", "")
    quality = result.get("quality", {})
    word_count = result.get("word_count") or quality.get("wordCount", 0)
    revisions = result.get("revisions", 0)
    issues = quality.get("issues", [])

    if args.no_write:
        print(final)
    else:
        out_path.write_text(final, encoding="utf-8")
        print(f"Wrote {out_path}")

    print(f"Done! {word_count} words, {revisions} revisions, ok={quality.get('ok', False)}")
    if issues:
        print("Quality issues:")
        for issue in issues:
            print(f"  - {issue}")
    return 0 if quality.get("ok", False) else 1


if __name__ == "__main__":
    raise SystemExit(main())
